xapian-applications/omega/index_file.cc

```diff
@@ -545,6 +545,77 @@ index_add_document(const string & urlterm, time_t last_altered,
     }
 }
 
+bool
+index_update_entry(const string & urlterm,
+		   DirectoryIterator & d,
+		   Xapian::docid & did)
+{
+    if (!did) {
+	Xapian::PostingIterator p = db.postlist_begin(urlterm);
+	if (p != db.postlist_end(urlterm))
+	    did = *p;
+    }
+
+    if (did) {
+	time_t mtime = d.get_mtime();
+	time_t ctime = d.get_ctime();
+	Xapian::Document doc = db.get_document(did);
+	string value_ctime = doc.get_value(VALUE_CTIME);
+	time_t doc_ctime = binary_string_to_int(value_ctime);
+	string value_mtime = doc.get_value(VALUE_LASTMOD);
+	time_t doc_mtime = binary_string_to_int(value_mtime);
+
+	if (doc_mtime == mtime && doc_ctime <= ctime) {
```

Comment:
While this looks correct to me, we can potentially optimise this check to avoid having to look up the values (or to only have to look up one in some cases).  This function potentially gets called for a lot of documents, so if we can eliminate many of them with a cheap check that's good.

We know upper and lower bounds on the values in a given slot, which translates to having bounds on the values of mtime that occur in the database (and similarly for ctime).  So for example we can cheaply check if the file's `mtime` is after any in the database, in which case we know that `doc_mtime == mtime` can't be true without having to even call `get_document()`.  Similarly we may also be able to tell that `doc_ctime <= ctime` must be true just from comparing `ctime` with the upper bound on ctime in the database.

Reply:
I am going to make a little research about it and try to implement it.